File: packages/unswallow/src/matrix.ts
import fs from 'node:fs';
import path from 'node:path';
import { matchesRange, parseRange } from './semver';
import type { EngineId, SwallowMatrixEntry, ToolPattern } from './types';

export interface MatrixFile {
  matrixVersion: string;
  updated: string;
  note?: string;
  entries: SwallowMatrixEntry[];
}

const CWD_MATRIX_PATH = path.join(process.cwd(), 'data', 'engine-matrix.json');

export function loadMatrix(custom?: SwallowMatrixEntry[]): SwallowMatrixEntry[] {
  if (custom && custom.length > 0) return custom;
  try {
    const pkg = require('@unswallow/matrix') as { entries: SwallowMatrixEntry[] };
    if (Array.isArray(pkg.entries)) return pkg.entries;
  } catch {
    // fall through to local paths (monorepo dev without workspace link)
  }
  for (const p of [CWD_MATRIX_PATH]) {
    try {
      const raw = fs.readFileSync(p, 'utf8');
      const file = JSON.parse(raw) as MatrixFile;
      if (Array.isArray(file.entries)) return file.entries;
    } catch {
      continue;
    }
  }
  return [];
}

export function getMatrixFile(custom?: SwallowMatrixEntry[]): MatrixFile {
  if (custom && custom.length > 0) {
    return { matrixVersion: 'custom', updated: 'n/a', entries: custom };
  }
  try {
    const pkg = require('@unswallow/matrix') as MatrixFile;
    if (pkg && Array.isArray(pkg.entries)) {
      return {
        matrixVersion: pkg.matrixVersion,
        updated: pkg.updated,
        entries: pkg.entries,
      };
    }
  } catch {
    // fall through
  }
  try {
    const raw = fs.readFileSync(CWD_MATRIX_PATH, 'utf8');
    return JSON.parse(raw) as MatrixFile;
  } catch {
    return { matrixVersion: 'none', updated: 'n/a', entries: [] };
  }
}

export function normalizeEngine(hint?: string): EngineId | 'unknown' {
  if (!hint) return 'unknown';
  const h = hint.trim().toLowerCase().replace(/[-_\s.]/g, '');
  if (h === 'vllm') return 'vllm';
  if (h === 'sglang') return 'sglang';
  if (h === 'llamacpp' || h === 'llama') return 'llama.cpp';
  return 'unknown';
}

function comparatorCount(range: string): number {
  const r = parseRange(range);
  if (!r) return 0;
  return r.comparators.reduce((n, group) => Math.max(n, group.length), 0);
}

export function matchMatrixEntry(
  entries: SwallowMatrixEntry[],
  engine: EngineId,
  version: string,
  pattern: ToolPattern
): SwallowMatrixEntry | null {
  const candidates = entries.filter(
    (e) =>
      e.engine === engine &&
      e.pattern === pattern &&
      matchesRange(version, e.versionRange)
  );
  if (candidates.length === 0) return null;
  candidates.sort(
    (a, b) => comparatorCount(b.versionRange) - comparatorCount(a.versionRange)
  );
  return candidates[0];
}

File: packages/python/unswallow/matrix.py
from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Dict, List, Optional

from .semver import matches_range, parse_range
from .types import EngineId, SwallowMatrixEntry, ToolPattern

BUNDLED_MATRIX = Path(__file__).parent / "data" / "engine-matrix.json"
ENV_MATRIX = os.environ.get("UNSWALLOW_MATRIX_PATH")


def load_matrix(custom: Optional[List[SwallowMatrixEntry]] = None) -> List[SwallowMatrixEntry]:
    if custom:
        return custom
    candidates = []
    if ENV_MATRIX:
        candidates.append(Path(ENV_MATRIX))
    candidates.append(BUNDLED_MATRIX)
    for p in candidates:
        try:
            raw = json.loads(p.read_text(encoding="utf-8"))
            entries = raw.get("entries", [])
            if isinstance(entries, list) and entries:
                return [SwallowMatrixEntry.from_json(row) for row in entries]
        except (OSError, ValueError):
            continue
    return []


def get_matrix_file(custom: Optional[List[SwallowMatrixEntry]] = None) -> Dict[str, object]:
    if custom:
        return {"matrixVersion": "custom", "updated": "n/a", "entries": custom}
    candidates = []
    if ENV_MATRIX:
        candidates.append(Path(ENV_MATRIX))
    candidates.append(BUNDLED_MATRIX)
    for p in candidates:
        try:
            raw = json.loads(p.read_text(encoding="utf-8"))
            raw["entries"] = [SwallowMatrixEntry.from_json(row) for row in raw.get("entries", [])]
            return raw
        except (OSError, ValueError):
            continue
    return {"matrixVersion": "none", "updated": "n/a", "entries": []}


def normalize_engine(hint: Optional[str]) -> str:
    if not hint:
        return "unknown"
    h = hint.strip().lower().replace("-", "").replace("_", "").replace(" ", "").replace(".", "")
    if h == "vllm":
        return "vllm"
    if h == "sglang":
        return "sglang"
    if h in ("llamacpp", "llama"):
        return "llama.cpp"
    return "unknown"


def _comparator_count(range_str: str) -> int:
    r = parse_range(range_str)
    if not r:
        return 0
    return max(len(group) for group in r)


def match_matrix_entry(
    entries: List[SwallowMatrixEntry],
    engine: EngineId,
    version: str,
    pattern: ToolPattern,
) -> Optional[SwallowMatrixEntry]:
    candidates = [
        e
        for e in entries
        if e.engine == engine and e.pattern == pattern and matches_range(version, e.version_range)
    ]
    if not candidates:
        return None
    candidates.sort(key=lambda e: _comparator_count(e.version_range), reverse=True)
    return candidates[0]
